eth/local_tx_handler.go

```diff
@@ -0,0 +1,497 @@
+		var txsToBroadcast []common.Hash
+
+		for addr := range l.txQueues {
+			if l.sentRecently(peerID(peer.ID()), addr) {
+				continue
+			}
+			if tx := l.nextTxToBroadcast(peer, addr); tx != nil {
+				txsToBroadcast = append(txsToBroadcast, *tx)
+			}
+		}
+		if len(txsToBroadcast) > 0 {
+			peer.AsyncSendTransactions(txsToBroadcast)
+		}
+	}
+}
+
+// maybeBroadcast announces the next unknown transaction to each peer if:
+//  1. the peer is in the square-root subset of the peerset
+//  2. another transaction from the same sender has not been announced to the peer recently
+func (l *localTxHandler) maybeAnnounce() {
+	allPeers := l.peers.allEthPeers()
+	allPeers = sortedPeers(allPeers)
+	announcePeers := allPeers[int(math.Sqrt(float64(len(allPeers)))):]
+	for _, peer := range announcePeers {
+		var txsToAnnounce []common.Hash
+
+		for addr := range l.txQueues {
+			if l.announcedRecently(peerID(peer.ID()), addr) {
+				continue
+			}
+			if tx := l.nextTxToAnnounce(peer, addr); tx != nil {
+				txsToAnnounce = append(txsToAnnounce, *tx)
+			}
+		}
+		if len(txsToAnnounce) > 0 {
+			peer.AsyncSendPooledTransactionHashes(txsToAnnounce)
+		}
+	}
+}
+
+// trimLocals removes transactions from monitoring queues of their senders
+// if their nonce is lte the sender account's nonce in the head state of the chain.
+func (l *localTxHandler) trimLocals() {
+	for addr, txs := range l.txQueues {
+		var cutPoint int
+		currentNonce := l.GetNonce(addr)
+		for i, tx := range txs {
+			if tx.Nonce() < currentNonce {
+				cutPoint = i
+			} else {
+				break
+			}
+		}
+
+		l.txQueues[addr] = l.txQueues[addr][cutPoint:]
+
+		if len(l.txQueues[addr]) == 0 {
+			delete(l.txQueues, addr)
+		}
+	}
+}
+
+// sortedPeers sorts a set of peers by the lexographic order of their rlpx secp256k1 public keys.
+// the provided array is not modified.
+func sortedPeers(unsorted []*ethPeer) []*ethPeer {
+	peerMap := make(map[string]*ethPeer)
+	peerMapKeys := []string{}
+
+	for _, peer := range unsorted {
+		pid := peer.ID()
+		peerMap[pid] = peer
+		peerMapKeys = append(peerMapKeys, pid)
+	}
+
+	sort.Strings(peerMapKeys)
+
+	var res []*ethPeer
+	for _, pid := range peerMapKeys {
+		res = append(res, peerMap[pid])
+	}
+
+	return res
+}
+
+// TODO: noting an edge-case/idea/question here.  If we have already broadcasted/announced a set of transactions to a peer and we receive one or more replacement transactions with the same nonce
+// should we broadcast/announce them all immediately?  seems like a no-brainer.
+
+// mergeNonceOrderedTxs combines curTxs and newTxs together.  each list is nonce-contiguous, nonce-unique ordered ascending by nonce.  If the nonce of a transaction in newTxs matches a transaction in curTxs, only the transaction from newTxs is included in the result.
+// The resulting list is nonce-contiguous, nonce-unique and ordered ascending by nonce.
+func mergeNonceOrderedTxs(curTxs []*types.Transaction, newTxs []*types.Transaction) []*types.Transaction {
```

Comment:
I got stuck here a bit, this seemed like the kind of thing that might be nicer to do with generics? Here's an untested proof of concept

```
func mergeNonceOrderedTxs(curTxs []*types.Transaction, newTxs []*types.Transaction) []*types.Transaction {
	cmpFn := func(a, b *types.Transaction) {
		if a.Nonce() < b.Nonce() {
			return -1
		}
		if a.Nonce > b.Nonce() {
			return 1
		}
		return 0
	}
	return mergeNonceOrderedGeneric(curTxs, newTxs, cmpFn)
}

func mergeNonceOrderedGeneric[T any](listA []T, listB []T, cmpFn func(a, b T) int) []T {
	// insert new txs into the sender's queue, keeping the resulting array nonce-ordered and
	// replacing pre-existing txs if there is a tx with same nonce
	var (
		res  []T
		idxA = 0
		idxB = 0
	)
	for {
		if idxA >= len(listA) || idxB >= len(listB) {
			break
		}
		txA, txB := listA[idxA], listB[idxB]
		switch cmp(txA, txB) {
		case -1:
			res = append(res, txA)
			idxA++
		case 0:
			idxA++
			fallthrough
		default:
			res = append(res, txB])
			idxB++
		}
	}
	// Merge is done, at least one of them is empty, just add whatever remains
	if idxA < len(listA) {
		res = append(res, listA[idxA:]...)
	}
	if idxB < len(listB) {
		res = append(res, listB[idxB:]...)
	}
	return res
}
```

I have a feeling it is cleaner that way: the sorting doesn't know about tx characteristics, the algo is fully generic.

Reply:
No need to make code generic unless there is an actual need to reuse it with multiple types.